Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h

```diff
@@ -746,32 +798,94 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
 
     size_t newByteOffset = thisObject->byteOffset() + offset * ViewClass::elementSize;
 
-    JSObject* defaultConstructor = globalObject->typedArrayConstructor(ViewClass::TypedArrayStorageType);
-    JSValue species = callFrame->uncheckedArgument(2);
-    if (species == defaultConstructor) {
+    scope.release();
+    return JSValue::encode(speciesConstruct(globalObject, thisObject, [&]() {
         Structure* structure = globalObject->typedArrayStructure(ViewClass::TypedArrayStorageType);
-
-        RELEASE_AND_RETURN(scope, JSValue::encode(ViewClass::create(
+        return ViewClass::create(
             globalObject, structure, WTFMove(arrayBuffer),
             thisObject->byteOffset() + offset * ViewClass::elementSize,
-            length)));
+            length);
+    }, [&](MarkedArgumentBuffer& args) {
+        args.append(vm.m_typedArrayController->toJS(globalObject, thisObject->globalObject(), arrayBuffer.get()));
+        args.append(jsNumber(newByteOffset));
+        args.append(jsNumber(length));
+        ASSERT(!args.hasOverflowed());
+    }));
+}
+
+template<typename ViewClass>
+static inline void validateIntegerIndex(JSGlobalObject* globalObject, ViewClass* thisObject, double index)
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+
+    if (UNLIKELY(thisObject->isDetached())) {
```

Comment:
Ah, I wish one day we would change the spec's [`IsValidIntegerIndex`](https://tc39.es/ecma262/#sec-isvalidintegerindex) not to include detached buffer check.

Reply:
I'm not sure there's any spec work left to be done there though; we stopped various operations from throwing, but given that the concept of a detached buffer exists, we do need to early out before attempting to interact with its contents.

The only redundancy I see is that integer-indexed ([[DefineOwnProperty]])[https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-defineownproperty-p-desc] ends up checking `IsValidIntegerIndex` twice because of the `IntegerIndexedElementSet` call.